From b85f41c1a41c34758a258b07edd82371c7b9cc19 Mon Sep 17 00:00:00 2001
From: Matthieu Mauger <matthieux.mauger@intel.com>
Date: Fri, 5 Jul 2013 10:46:39 +0200
Subject: [PORT FROM MAIN] P2P: Inform upper layer when WFD-IE field of a P2P device has changed

BZ: 118015

When WFD IE of a P2P device changes, the field is locally updated in P2P
supplicant but upper layer is not informed about this change.

This patch fixes that.

Change-Id: Ie6cd92717d6e760e92b4fcd76a8895b95b4e791a
Orig-Change-Id: I7ae84704108cdb640ffee617fc5b26153ea42ac0
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Matthieu Mauger <matthieux.mauger@intel.com>
---
 src/p2p/p2p.c |   21 ++++++++++++++++++++-
 1 files changed, 20 insertions(+), 1 deletions(-)

diff --git a/src/p2p/p2p.c b/src/p2p/p2p.c
index a3a83bf..ea60aba 100644
--- a/src/p2p/p2p.c
+++ b/src/p2p/p2p.c
@@ -625,6 +625,22 @@ static void p2p_copy_wps_info(struct p2p_data *p2p, struct p2p_device *dev,
 }
 
 
+static int p2p_compare_wfd_info(struct p2p_device *dev,
+			      const struct p2p_message *msg)
+{
+	if (dev->info.wfd_subelems && msg->wfd_subelems) {
+		if (dev->info.wfd_subelems->used != msg->wfd_subelems->used)
+			return 1;
+
+		return memcmp(dev->info.wfd_subelems->buf, msg->wfd_subelems->buf,
+				  dev->info.wfd_subelems->used);
+	}
+	if (dev->info.wfd_subelems || msg->wfd_subelems)
+		return 1;
+
+	return 0;
+}
+
 /**
  * p2p_add_device - Add peer entries based on scan results or P2P frames
  * @p2p: P2P module context from p2p_init()
@@ -650,6 +666,7 @@ int p2p_add_device(struct p2p_data *p2p, const u8 *addr, int freq,
 	struct p2p_device *dev;
 	struct p2p_message msg;
 	const u8 *p2p_dev_addr;
+	int wfd_changed;
 	int i;
 	struct os_time time_now;
 
@@ -761,6 +778,8 @@ int p2p_add_device(struct p2p_data *p2p, const u8 *addr, int freq,
 			break;
 	}
 
+	wfd_changed = p2p_compare_wfd_info(dev, &msg);
+
 	if (msg.wfd_subelems) {
 		wpabuf_free(dev->info.wfd_subelems);
 		dev->info.wfd_subelems = wpabuf_dup(msg.wfd_subelems);
@@ -776,7 +795,7 @@ int p2p_add_device(struct p2p_data *p2p, const u8 *addr, int freq,
 	if (p2p_pending_sd_req(p2p, dev))
 		dev->flags |= P2P_DEV_SD_SCHEDULE;
 
-	if (dev->flags & P2P_DEV_REPORTED)
+	if (dev->flags & P2P_DEV_REPORTED && !wfd_changed)
 		return 0;
 
 	p2p_dbg(p2p, "Peer found with Listen frequency %d MHz (rx_time=%u.%06u)",
-- 
1.7.4.1

